sklearn/feature_selection/tests/test_sequential.py

```diff
@@ -39,9 +40,11 @@ def test_n_features_to_select(direction, n_features_to_select):
     sfs.fit(X, y)
     if n_features_to_select is None:
         n_features_to_select = 5  # n_features // 2
-    assert sfs.get_support(indices=True).shape[0] == n_features_to_select
-    assert sfs.n_features_to_select_ == n_features_to_select
-    assert sfs.transform(X).shape[1] == n_features_to_select
+
+    if aborted_rate is None:
+        assert sfs.get_support(indices=True).shape[0] == n_features_to_select
+        assert sfs.n_features_to_select_ == n_features_to_select
+        assert sfs.transform(X).shape[1] == n_features_to_select
 
 
```

Comment:
if `tol=1e-3` and `n_features_to_select=X.shape[1]` you can check that the number of selected features is lower than `X.shape[1]`.

You can also check that the cross-validation score of a model fitted on a strict random subset of the selected feature is lower than the mean cross-validation score of the final model by more than `tol`.

Maybe I would put this in a dedicated new test named `test_stopping_criterion`.

Reply:
And you can write a similar (dual) test `test_backward_stopping_criterion` for the backward strategy.

Reply:
Maybe it would be easier to write 2 independent tests for the 2 directions.

In both tests, we should check that when using `tol=1e-3` and `n_features_to_select=X.shape[1]` the final number of selected features `selector.support_.sum()` is strictly smaller than `X.shape[1]`.

Furthermore the cross-validation score of the model without feature selection should be worse than the CV score of the model with feature selection (without using `tol` in this comparison).

Then for the forward test, add the following assertion:

- select a new features at random among those not currently selected in `selector.support_`, build a new version of the data that includes all the features in `selector.support_` + this newly selected feature and check that the cross-validation score of the model trained on this new dataset variant is lower than the model with the selected forward selected features or at least does not improve by more than the tol margin.

Then for the backward direction do the opposite check: instead of adding a new feature to `selector.support_`, try to remove one of those selected features at random and check that the cross-validation score is either decreasing or not improving by more than the tol margin.

Reply:
added test for score comparison on https://github.com/scikit-learn/scikit-learn/pull/20145/commits/05e86300a864d1e8f5cd439630143fa2d1d0ecc5, https://github.com/scikit-learn/scikit-learn/pull/20145/commits/2f12708e2fe0432dddb0df2d224f02a1ef5d6dc1 and https://github.com/scikit-learn/scikit-learn/pull/20145/commits/f047e05c91abd2c5bb2dd2633ab203b8f3052c26.

And, added test for checking num of features on https://github.com/scikit-learn/scikit-learn/pull/20145/commits/2640d7873cbf6400f3bf2d7e1d6d31f165e4f2c4.